### DISCUSSION TITLE
[Guide] How to send Youtube, Twitch.. to mpv from your browser

### DISCUSSION THREAD

**Discussion by gunir** [mpv-player/mpv#14506] [Show and tell]
Original Thread: https://github.com/mpv-player/mpv/issues/9364

For people who curious, this is a plugin that allows the use of `mpv://` protocol: [protocol_hook.lua](https://github.com/FirefoxUniverse/FirefoxTweaksVN/blob/main/mpv/scripts/protocol_hook.lua)

You need to install this script to your Userscript Manager (preferbly Violentmoney): [Handlers Helper.user.js](https://github.com/FirefoxUniverse/FirefoxTweaksVN/blob/main/userscript/Handlers%20Helper.user.js)

And patch your browser handlers.json (`about:support` -> `Open Profile Folder`) using this (Paste the content of your `handlers.json` to the textarea -> Copy -> Paste back to your `handlers.json`: [handlers.json Patcher](https://htmlpreview.github.io/?https://github.com/FirefoxUniverse/FirefoxTweaksVN/blob/main/utility/handlers_patcher.html)

For advanced usage, install m3u8.user.js to enable watching from websites that yt-dlp doesn't support: [m3u8.user.js](https://github.com/FirefoxUniverse/FirefoxTweaksVN/blob/main/userscript/m3u8.user.js)

If you do everything correctly, congrats, this is how you use it: [DEMO VIDEO](https://streamable.com/4s1gfi)

Absolutely game changer, no more copying pasting, just drag and drop.

Enjoy!

**Comment by anatasiajp** [comment]
Bookmark

**Comment by ahaoboy** [comment]
tampermonkey script:
https://github.com/LuckyPuppy514/Play-With-MPV

Use shortcut key(default ctrl+v) to play video from clipboard, supports file, webdav and jellyfin.
https://github.com/mpv-easy/mpv-easy/tree/main/mpv-clip-play

**Reply by gunir** [reply]
Is there a English guide for Play With MPV, honestly I can't understand one bit, it requires to install mpv handler to system wide just like this thread (using Firefox's handlers.json instead, which is non-system wide and pretty safe, as said by mpv dev https://github.com/mpv-player/mpv/issues/9364#issuecomment-953296324) ?

**Reply by ahaoboy** [reply]
Unfortunately it seems to be only in Chinese
It is difficult to play web videos using a local player. The browser has your login information and verification information. Generally, you can only get links such as m3u that can be played directly in the browser and then open the video through a custom protocol or calling a local script.

This is why mpv-easy uses copy-paste implementation. Although it still needs to rely on the clipboard, and copy the URL, open the player, paste the URL and other steps, it seems to be the simplest implementation.

I am also looking for a simple and convenient way to play remote videos. Ideally, users should only install the mpv player and Chrome or tampermonkey script. Adding configuration and executing programs is very difficult for many ordinary users.

**Reply by gunir** [reply]
Then I probably prefer this thread method, it's working for me and it's safer as it doesn't poison or expose system handlers to all apps, only my browser.

**Comment by ahaoboy** [comment]
For those who want to support this feature in their own programs, here are the simple steps

## install protocol

create mpv.reg file
```
Windows Registry Editor Version 5.00
[HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Google\Chrome]
"ExternalProtocolDialogShowAlwaysOpenCheckbox"=dword:00000001

[HKEY_LOCAL_MACHINE\SOFTWARE\Policies\Microsoft\Edge]
"ExternalProtocolDialogShowAlwaysOpenCheckbox"=dword:00000001

[HKEY_CLASSES_ROOT\mpv-easy]
@="mpv-easy"
"URL Protocol"=""

[HKEY_CLASSES_ROOT\mpv-easy\DefaultIcon]
@=""

[HKEY_CLASSES_ROOT\mpv-easy\shell]
@=""

[HKEY_CLASSES_ROOT\mpv-easy\shell\open]
@=""

[HKEY_CLASSES_ROOT\mpv-easy\shell\open\command]
@="C:\\app\\mpv-test\\mpv.exe %1"
```

install protocol, note that you must use Windows-style separators
```
cmd /c "regedit.exe /S C:\test\mpv.reg"
```
Now when you open the link mpv-easy:abcd through the browser, the browser will pop up the following prompt message
![image](https://github.com/user-attachments/assets/aefa8383-e775-422d-b411-da5b2500c8d0)

## send message to mpv
When you use mpv-easy://abcd to start the mpv player, the parameter actually received by the mpv program is mpv-easy://abcd
You can use the following complex PowerShell script to remove the protocol header mpv-easy:// and use PowerShell's UrlDecode function to decode the link
```
[HKEY_CLASSES_ROOT\mpv-easy\shell\open\command]

@="C:\\Windows\\System32\\WindowsPowerShell\\v1.0\\powershell.exe -WindowStyle Hidden -Command \"& {Add-Type -AssemblyName System.Web;$PARAMS=([System.Web.HTTPUtility]::UrlDecode('%1') -replace '^mpv-easy://'); Start-Process -FilePath \\\"C:\\app\\mpv-test\\mpv.exe\\\" -ArgumentList $PARAMS}\""
```
After reinstalling the reg file, you can use link to play the video
```
mpv-easy://c://xxx.mp4 
mpv-easy://https://www.youtube.com/watch?v=xxx
```

## Some optimizations
Although we can use encodeURIComponent to pass arbitrary parameters to mpv on the web, there are still some problems

### max length
The maximum length of a windows URL is 2048, which is sufficient for most scenarios.

### program flicker
Starting mpv via powershell, there will be a very short program flicker

### playlist
mpv only supports playlist via m3u files

In order to solve the above problems, we can implement a transfer program
```
[HKEY_CLASSES_ROOT\\mpv-easy\\shell\\open\\command]
@="${mpvPlayWithPath} ${mpvPath} %1"
```
Compress the data using gzip on the web, then convert it to base64 encoding, which can solve any encoding problems and reduce the link length
If the decoded video is a playlist, you can create a temporary m3u file in the transfer program, and then use the --playlist parameter to play multiple videos in mpv.


I used the above ideas to implement [mpv-easy-play-with](https://github.com/mpv-easy/mpv-easy/tree/main/mpv-play-with)
If you have better ideas, please feel free to share.

**Comment by iamthelara** [comment]
""""And patch your browser handlers.json (about:support -> Open Profile Folder) using this (Paste the content of your handlers.json to the textarea -> Copy -> Paste back to your handlers.json: [handlers.json Patcher](https://htmlpreview.github.io/?https://github.com/FirefoxUniverse/FirefoxTweaksVN/blob/main/utility/handlers_patcher.html)""""

I could not understand this part. Would you please elaborate? @gunir

**Comment by z3us** [comment]
I don´t understand .
How can I patch handlers.json (chrome)? I can´t find that file

**Comment by Obegg** [comment]
I did not try it yet, but it looks awesome by the looks of the video.
Just wanted to share that there are more ways to do that, here's an example - https://github.com/mpv-player/mpv/discussions/15565
Not sure if it will be the best approach but I thought I should at least mention it since it wishes to accomplish the same goal as this discussion.

**Comment by ahaoboy** [comment]
There are two ways to play remote video

The first method is to install a browser plugin and register a local protocol, and call the local player through the plugin in the browser
https://github.com/mpv-easy/mpv-easy/tree/main/mpv-play-with

https://github.com/user-attachments/assets/80aa6659-bd60-464a-b3c4-4df8d1dd5e8a

The second is to copy the browser link and paste it into the local player
https://github.com/mpv-easy/mpv-easy/tree/main/mpv-clipboard-play

https://github.com/user-attachments/assets/369dda83-6ea4-4bdb-86ec-21d78d92f5e3

**Reply by Obegg** [reply]
Have you seen the post above yours? There's a third way, you missed it
